### DISCUSSION TITLE
Primitives issue

### DISCUSSION THREAD

**Discussion by VedankNaik** [iTwin/itwinjs-core#4113] [Q&A]
Hello, we have implemented [Simple 3d](https://www.itwinjs.org/sandboxes/iTwinPlatform/Simple%203d). This works fine for some models but for certain models primitives are rendered as weird shapes. 
Image attached below is sphere. 

![image](https://user-images.githubusercontent.com/47364632/185081615-a0dbb756-7541-4b5a-9135-8afa8ed2024d.png)

What could be the issue?

Thanks.

**Comment by pmconne** [comment]
Show your code.

**Reply by VedankNaik** [reply]
Code is same as shown in Simple 3d sample. Is there any specific snippet you would like to see?

**Reply by pmconne** [reply]
I would like you to explain how to reproduce the image you showed above.
I followed your link, selected "Sphere" from the combo box, and observed a sphere.
Please fill in the missing steps.

**Reply by VedankNaik** [reply]
I was not able to reproduce this in sandbox. Only thing I changed was model where this issue appears. Can model size cause issue? Because it works fine on sample models, we noticed this only on large scale model.

**Reply by pmconne** [reply]
I have explained at length on this forum the problems with that Simple 3d example, but for the life of me I cannot find that discussion now. The example does some dumb things:
1. It creates the decoration graphics once and then reuses them every frame. The level of detail of a graphic is computed based on the size of the geometry in the viewport at the time the graphic is created. If you subsequently zoom in, the level of detail will be too low, and you'll observe low LOD graphics like in your screenshot.

You should instead recreate your graphics every time your decorator's `decorate` method is invoked. You should generally override [useCachedDecorations](https://www.itwinjs.org/reference/core-frontend/views/viewportdecorator/usecacheddecorations/) to be `true`.

2. It uses a PolyfaceBuilder to turn a sphere (or box, or torus, or whatever) into a mesh, then passes the mesh to the GraphicBuilder. The mesh is going to have some level of detail of its own and it will probably be wrong.

You should instead pass the sphere (or other solid primitive) directly to [GraphicBuilder.addSolidPrimitive](https://www.itwinjs.org/reference/core-frontend/rendering/graphicbuilder/#addsolidprimitive).

@Ellord207 this problem keeps coming up. Can we fix the GeometryDecorator to serve as a better example?

**Reply by Ellord207** [reply]
I have had my own discussion with @Josh-Schifter on this topic too.  He was working on improvements to the geometry samples, which included resolving issue 1 in the GeometryDecorator.  I'll check the status of that change and see if there we can resolve issue 2 at the same time.

**Reply by VedankNaik** [reply]
> You should instead pass the sphere (or other solid primitive) directly to [GraphicBuilder.addSolidPrimitive](https://www.itwinjs.org/reference/core-frontend/rendering/graphicbuilder/#addsolidprimitive).

Can you give more pointers on this?

**Reply by pmconne** [reply]
[Example](https://github.com/iTwin/itwinjs-core/blob/2048398a4c4dde16e92661259c856831690996b3/test-apps/display-test-app/src/frontend/DecorationGeometryExample.ts#L29).
